From 2994abbd523ae0cef2e3d43a43460e2f7e54373d Mon Sep 17 00:00:00 2001
From: Ryan O'Hara <rohara@redhat.com>
Date: Mon, 13 Jan 2014 10:37:18 -0600
Subject: [PATCH] Ignore EINTR from sem_timedwait

Signed-off-by: Ryan O'Hara <rohara@redhat.com>
---
 pulse.c | 18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/pulse.c b/pulse.c
index a9cd4ab..0fd317d 100644
--- a/pulse.c
+++ b/pulse.c
@@ -1432,12 +1432,18 @@ activateLvs (struct lvsConfig *config, int flags)
 	  execv (lvsBinary, argv);
 	  exit (-1);
       }
-      if (sem_timedwait (sem, &ts) != 0) {
-	  piranha_log (flags, (char *) "Error waiting for semaphore: %s", strerror(errno));
-	  sem_unlink (LVS_MUTEX);
-	  kill (child, SIGKILL);
-	  exit (-1);
-      }
+
+      do {
+          if (sem_timedwait (sem, &ts) != 0) {
+              if (errno == EINTR)
+                  continue;
+              piranha_log (flags, (char *) "Error waiting for semaphore: %s", strerror(errno));
+              sem_unlink (LVS_MUTEX);
+              kill (child, SIGKILL);
+              exit (-1);
+          }
+      } while (0);
+
       sem_unlink (LVS_MUTEX);
     }
 
-- 
1.8.1.4

