From bf84476259eb1ff0c614fb67af202345149d572d Mon Sep 17 00:00:00 2001
From: Ryan O'Hara <rohara@redhat.com>
Date: Thu, 7 Apr 2016 09:55:06 -0500
Subject: [PATCH] Set remote port prior to comparison.

Resolves: #1319376

Signed-off-by: Ryan O'Hara <rohara@redhat.com>
---
 lvsd.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/lvsd.c b/lvsd.c
index 10d7749..d99865d 100644
--- a/lvsd.c
+++ b/lvsd.c
@@ -291,8 +291,9 @@ findClientConfig (struct lvsConfig *config, int *vserver, int *service,
 	  tinfo.vip = config->virtServers[vsn].virtualAddress;
 	  tinfo.address = config->virtServers[vsn].servers[sn].address;
 	  tinfo.protocol = config->virtServers[vsn].protocol;
-	  tinfo.port = config->virtServers[vsn].servers[sn].port;
-	  
+	  tinfo.port = config->virtServers[vsn].port;
+	  tinfo.rport = config->virtServers[vsn].servers[sn].port;
+
 	  if (cmpRealServers(&tinfo, needle))
 	    continue;
 
-- 
1.8.1.4

