From 030e466070df88f4ed3573060a873b7231efb794 Mon Sep 17 00:00:00 2001
From: Marek 'marx' Grac <mgrac@redhat.com>
Date: Wed, 17 Feb 2010 13:30:08 +0100
Subject: [PATCH] Fix sorry server.

Resolves: #903711

Signed-off-by: Ryan O'Hara <rohara@redhat.com>
---
 lvsd.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/lvsd.c b/lvsd.c
index b718bab..10a25ca 100644
--- a/lvsd.c
+++ b/lvsd.c
@@ -145,11 +145,11 @@ struct clientInfo
 
 int
 shutdownDev (int flags, char *ipvsadm, char *virtualAddress, int port,
-	     struct in_addr *remoteAddr, int service_type, int fm, int udp);
+	     struct in_addr *remoteAddr, int rport, int service_type, int fm, int udp);
 
 int
 bringUpDev (int flags, char *ipvsadm, char *virtualAddress, int port,
-	    struct in_addr *remoteAddr, char *routingMethod, int weight,
+	    struct in_addr *remoteAddr, int rport, char *routingMethod, int weight,
 	    int service_type, int fm, int udp);
 
 char *
@@ -1194,6 +1194,7 @@ checkClientMonitor(struct lvsConfig *config, struct lvsVirtualServer *vs,
 	       config->vsadm,
 	       buf, vs->port,
 	       &vs->sorry_server,
+	       vs->port,
 	       rm_to_str(config->redirectType),
 	       1, SERV_LVS,
 	       vs->fwmark, (vs->protocol == IPPROTO_UDP));
@@ -1210,6 +1211,7 @@ checkClientMonitor(struct lvsConfig *config, struct lvsVirtualServer *vs,
 	        config->vsadm,
 	        buf, vs->port,
 	        &vs->sorry_server,
+	        vs->port,
 	        SERV_LVS,
 	        vs->fwmark, (vs->protocol == IPPROTO_UDP));
   }
-- 
1.8.1

